### ISSUE TITLE
SwiftUI support

### ISSUE THREAD

**Issue by djbe** [SwiftGen/SwiftGen#750]
Seeing as SwiftUI is reaching the point where it's becoming generally usable with iOS 14, we're going to need to adapt our templates to support it.

Users have been asking about this:
- Assets support: #666 
- Fonts support: #679
- Strings support: #685
- Color template suggestion in #747

### Asset Catalog

`Color` instead of `NS/UIColor`, `Image` instead of `NSImage/UIImage`, etc...

### Fonts

Uses a `Font` type instead of `NS/UIFont`.

## Strings

This one's a doozy, the system is quite different. The localisation key format is totally different, where parameter placeholders **must** be in the key, but they can be anywhere in the key. We return `LocalizedStringKey` objects instead of `String`, that contain both the key and the filled in parameters.

## SwiftGen Templates

Do we adapt existing templates to provide SwiftUI compatible properties? Or do we create separate SwiftUI templates? How do we handle different swift versions?

What about mixed projects (with both SwiftUI and AppKit/UIKit code)? Can we use SwiftUI types from UIKit? Or vice versa?

----

Once we decide on these topics, I'll open a dedicated issue for each template to discuss the implementation, before creating a PR.

**Comment by avi-screenovate** [comment]
Has there been any movement on SwiftUI support?

**Comment by djbe** [comment]
I'd love to get some traction for this, but as I'm not using SwiftUI yet myself, I don't really know what's needed and/or what works best. 

But a more important issue is deciding how we handle dedicated templates for for example SwiftUI, or [for newer Swift versions](#748), or if we put everything in one (complicated) template.

Another issue is can (and should?) we handle the use case for **mixed** SwiftUI/UIKit projects? I already find it annoying that we need to have separate Obj-C templates (only for strings now)... So, like I mentioned in my first post, we need to decide on how to handle these things before we can even start on the templates.

The good thing is, once we know the answer to these, writing the templates is **easy** 😄

**Comment by DenTelezhkin** [comment]
Posting this here for visibility, been using this approach for quite some time now, feels great and native to SwiftUI https://github.com/SwiftGen/SwiftGen/issues/666#issuecomment-625341873.

**Comment by avi-screenovate** [comment]
I think that having a `Color` property along with `UI/NSColor` and same for image would be a great start.

**Comment by djbe** [comment]
Let's keep the parser/template specific discussion in their specific issue(s).

The discussion here is to try and decide a general direction on how to best handle these (and future) changes, without having to resort to "quick fixes", valid as those may be. What I mean with that is that I'd like to work out a concrete plan, which'll probably consist of a few evolutionary steps, before implementing anything. A first step **could** be just adding some properties.

**Comment by Jeehut** [comment]
@djbe I'm currently developing my first bigger application using pure SwiftUI and I'd love to have a dedicated template for SwiftUI. In a another project we're starting to use SwiftUI as well, but there we need both, support for SwiftUI and UIKit. I think many new projects will opt for pure SwiftUI already (and even more so in the future), but many projects will also need to support both.

So my personal suggestion would be to **adjust the current UIKit templates** in a way that they don't interfere with SwiftUI (for example the fonts template `swift5` overrides the SwiftUI `Font` type and therefore can't be used) and provide **new templates for SwiftUI only**. All the SwiftUI APIs are new and therefore it should be easy to do this separation. Then, we should point users who want to use both UIKit and SwiftUI in a project to just use two templates for each feature, suffixing the files with `-UIKit` and `-SwiftUI` respectively. This should also make it a little easier to write the templates as they are more focused.

That's what I think would solve all the use cases I can think of.

**Comment by Jeehut** [comment]
FYI: As part of my first pure SwiftUI app, I created custom templates for assets, strings and fonts which work well for me.

I posted my templates and with some more details in the respective issues, here's an overview for your convenience:

- [My XCAssets custom template](https://github.com/SwiftGen/SwiftGen/issues/666#issuecomment-782894714)
- [My Fonts custom template](https://github.com/SwiftGen/SwiftGen/issues/679#issuecomment-782896985)
- [My Strings custom template](https://github.com/SwiftGen/SwiftGen/issues/685#issuecomment-782893242)

I hope it helps.

**Comment by sisoje** [comment]
one year later... is swiftgen officially a deadware?

**Comment by carmelogallo** [comment]
Any news on this one?

**Comment by davidscheutz** [comment]
@djbe where is the traction? 🔍

**Comment by wangpeiyan** [comment]
is there any progress here

**Comment by Jeehut** [comment]
For the Strings files in particular, I have developed an app that generates enums much like SwiftGen with full SwiftUI support, it's currently in Beta, check it out here: https://github.com/FlineDev/ReMafoX/releases

This is what the output of ReMafoX looks like (extract from [this file](https://github.com/FlineDevPublic/OpenFocusTimer/blob/main/Sources/Resources/ReMafoX.swift)):

```swift
import Foundation
import SwiftUI

/// Top-level shortcut for ``Res.Str``. Provides safe access to localized strings. Managed by ReMafoX (https://remafox.app).
public typealias Loc = Res.Str

/// Top-level namespace for safe resource access. Managed by ReMafoX (https://remafox.app).
public enum Res {
   /// Root namespace for safe access to localized strings. Managed by ReMafoX (https://remafox.app).
   public enum Str {
      public enum Main {
         public enum Detail {
            /// 🇺🇸 English: "No sidebar entry selected."
            public enum EmptyState {
               /// The translated `String` instance.
               public static var string: String { Bundle.module.localizedString(forKey: self.tableLookupKey, value: nil, table: "Localizable") }

               /// The SwiftUI `LocalizedStringKey` instance.
               public static var locStringKey: LocalizedStringKey { LocalizedStringKey(self.tableLookupKey) }

               /// The lookup key in the translation table (= the key in the `.strings` or `.stringsdict` file).
               public static var tableLookupKey: String { "main.detail.empty_state" }
            }
         }
      }
   }
}
```

**Comment by peterpaulis** [comment]
https://www.raywenderlich.com/23709326-swiftgen-tutorial-for-ios
